<a href="https://colab.research.google.com/github/Hertie-School-Machine-Learning-F2022/git-github-assignment-and-fundamentals-nikpaw/blob/main/problem_set_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem set 1

## Student name: Niklas Pawelzik

## Submit by: Thursday 22.09.2022, 23:59

#### You will go over some of the things that we saw in class today, and will answer the quiz based on the outcome of your computational process. You also need to submit the script to demostrate the process you used to get to the answers.

For this problem set you DO NOT NEED to set training and testing data.

## Question 1.

In [ ]:
#from sklearn.datasets import load_diabetes and set X and y
from sklearn import datasets
X,y = datasets.load_diabetes(return_X_y = True, as_frame = True)

X
 

# Take some time to understand the dataset
# take a look at it as a DataFrame object


# For this problem DO NOT SET training and testing data. We will use the same data
# throughout the problem set.

age       sex       bmi        bp        s1        s2        s3  \
0    0.038076  0.050680  0.061696  0.021872 -0.044223 -0.034821 -0.043401   
1   -0.001882 -0.044642 -0.051474 -0.026328 -0.008449 -0.019163  0.074412   
2    0.085299  0.050680  0.044451 -0.005671 -0.045599 -0.034194 -0.032356   
3   -0.089063 -0.044642 -0.011595 -0.036656  0.012191  0.024991 -0.036038   
4    0.005383 -0.044642 -0.036385  0.021872  0.003935  0.015596  0.008142   
..        ...       ...       ...       ...       ...       ...       ...   
437  0.041708  0.050680  0.019662  0.059744 -0.005697 -0.002566 -0.028674   
438 -0.005515  0.050680 -0.015906 -0.067642  0.049341  0.079165 -0.028674   
439  0.041708  0.050680 -0.015906  0.017282 -0.037344 -0.013840 -0.024993   
440 -0.045472 -0.044642  0.039062  0.001215  0.016318  0.015283 -0.028674   
441 -0.045472 -0.044642 -0.073030 -0.081414  0.083740  0.027809  0.173816   

           s4        s5        s6  
0   -0.002592  0.019908 -0.017646  
1   -0.039493 -0.068330 -0.092204  
2   -0.002592  0.002864 -0.025930  
3    0.034309  0.022692 -0.009362  
4   -0.002592 -0.031991 -0.046641  
..        ...       ...       ...  
437 -0.002592  0.031193  0.007207  
438  0.034309 -0.018118  0.044485  
439 -0.011080 -0.046879  0.015491  
440  0.026560  0.044528 -0.025930  
441 -0.039493 -0.004220  0.003064  

[442 rows x 10 columns]

How many variables are in X?

**Answer:**

*Variable number is equal to column number, thus there are ten variables.*

## Question 2

In [ ]:
# Import the SGD Linear Regression algorithm 
# SGD stands for Stochastic Gradient Descent: the gradient of the loss is 
# estimated each sample at a time and the model is updated along the way 
# with a decreasing strength schedule (aka learning rate).
from sklearn.linear_model import SGDRegressor

# Set it as the mod variable
mod = SGDRegressor()

# Take a look at the parameters
mod.get_params()

{'alpha': 0.0001,
 'average': False,
 'early_stopping': False,
 'epsilon': 0.1,
 'eta0': 0.01,
 'fit_intercept': True,
 'l1_ratio': 0.15,
 'learning_rate': 'invscaling',
 'loss': 'squared_error',
 'max_iter': 1000,
 'n_iter_no_change': 5,
 'penalty': 'l2',
 'power_t': 0.25,
 'random_state': None,
 'shuffle': True,
 'tol': 0.001,
 'validation_fraction': 0.1,
 'verbose': 0,
 'warm_start': False}

What are the parameters of the Linear Regression model?

**Answer:** 

*'epsilon', 'learning_rate' and 'penalty' are arguments of SGDRegressor. 'normalize' is not an argument of SGDRegressor.*

## Question 3.

In [ ]:
# Import the Standard Scaler and take a look at the parameters
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.get_params()


{'copy': True, 'with_mean': True, 'with_std': True}

What are the default option at which with_mean and with_std are set?

**Answer:** 

*The default option of with_mean is set to default=True.*

*The default option of with_std is set to default=True*

## Question 4.

In [ ]:
# Create a pipeline using the two previous steps, for the model
# set the SGDRegressor parameter random_state = 123 to reproduce results.

from sklearn.pipeline import Pipeline

pipe = Pipeline([
    ('scaler', StandardScaler()),     
    ('model', SGDRegressor(random_state = 123))
    ])

#Fit and predict the pipe, print results

pred = pipe.fit(X,y).predict(X)

pred

array([203.33036868,  71.44132129, 174.18891477, 162.71569477,
       128.16421117, 106.46855937,  78.15730757, 121.62004635,
       158.74826584, 213.76124061, 100.98211756, 103.73467041,
       115.03992428, 161.20630252, 102.6303906 , 171.97711381,
       209.73110688, 180.81360592, 146.07389902, 122.1424965 ,
       117.92285927,  90.98734606, 118.17218061, 274.4655061 ,
       163.9397713 , 143.2871305 ,  96.67844807, 177.95194312,
       127.41767326, 183.54905596, 159.43389974,  69.77413097,
       255.02234683, 108.1211581 ,  81.21817834,  83.17022348,
       207.09177574, 154.32172729, 241.77566415, 134.31684727,
       151.34086198,  73.70253571, 142.29831916,  79.3959324 ,
       216.42832247, 123.27110149, 140.07339717, 106.78414486,
        77.56835256, 184.29213326, 157.61488977, 166.75142807,
       133.16057603, 157.67450941, 141.39374227,  75.07925504,
       204.50631377,  80.93999369,  97.49830527, 132.8776951 ,
       115.85130194, 173.87422053,  67.65618223, 100.03

What are the first 4 predicted values? 

**Answer:**

*The first four values are 203.33036868,  71.44132129, 174.18891477 and 162.71569477.*

Question 5.

In [ ]:
# Now create a GridSearchCV Object using the previous pipe object,
# where the parameter 'model__alpa' is set to try 0.0001, 0.005, 0.001.
# Set cross validation to 3 times.

from sklearn.model_selection import GridSearchCV

gsv = GridSearchCV(estimator = pipe,
                 param_grid = {
                     'model__alpha': [0.0001, 0.005, 0.001],          
                 },
                 cv=3)




How are you supposed to write the values of the parameter 'alpha'?

**Answer:**

*Parameter 'alpha' requires a list of numbers, therefore we use  '[ ]' and no further quotation marks.*

Question 6.

In [ ]:
# Now fit the GridSearch model from above and see the results 
# (Hint: to make sure we all get the same results, you had to use the pipe that
# was set in question 4 making sure we used the same random_state = 123)
import pandas as pd

gsv.fit(X, y)

pd.DataFrame(gsv.cv_results_)



mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.007710      0.001536         0.007752        0.007251   
1       0.005511      0.000288         0.001907        0.000135   
2       0.005844      0.000821         0.001837        0.000108   

  param_model__alpha                    params  split0_test_score  \
0             0.0001  {'model__alpha': 0.0001}           0.466105   
1              0.005   {'model__alpha': 0.005}           0.465869   
2              0.001   {'model__alpha': 0.001}           0.466062   

   split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0           0.489906           0.502306         0.486106        0.015021   
1           0.490301           0.502853         0.486341        0.015356   
2           0.489980           0.502408         0.486150        0.015083   

   rank_test_score  
0                3  
1                1  
2                2

Which is the best model? Look at the rank_test_value

**Answer:**

*The rank_test_score proves model 1 (alpha = 0.005) to be the best model.*

Question 7.

Find out which test score is the one that is reported above in mean_test_score. What is it and how do you interpret this results?

**Answer:**

*R2 is relevant for regression models, accuracy is a measurement used to evaluate classification models. In the given case, we use mean_test_score to evaluate the models, it is a R2-based score. As a R2-based score, the mean_test_score indicates how well a model fits the data. Generally speaking, a high value is preferable over a low value; however, goodness of fit is not the only quality criterion of a regression model. An extremely high fitting score could also indicate an overfit of the model, meaning that the model that performs great on the training data at the cost of performing worse on real world or test data.*